index.js

```diff
@@ -128,6 +122,24 @@ class TuyaDevice extends EventEmitter {
       try {
         // Send request
         this._send(buffer).then(async data => {
+          if (data === 'json obj data unvalid') {
+            // Some devices don't respond to DP_QUERY so, for DPS get commands, fall
+            // back to using SEND with null value (so far, this appears to always work)
+            // For schema there's currently no fallback so provide a clear response
+            if (options.schema === true) {
+              data = 'Schema for device not available';
```

Comment:
Instead of returning values in cases of an error, we should be throwing errors instead.

Reply:
OK, but as a consumer, how do I associate the null response I receive with an error?  I mean, the get() call failed, it's not a null response from the device, it's a failure to get a response.  As a consumer of get() how can I know that the error emitted was the get call failed for this device?  I mean, I guess we could define an error strucure, with deviceId and request type so that the consumer would have some way to know that a given error was caused by a specific request.

However, my breaking API thought was to just make the response to get itself a JSON structure with a "response" property that contains the response and an optional "error" property if there is an error.  If we don't want to break existing we could add some type of flag to the get call to enable JSON responses.

Reply:
It might be more clear if we define what should happen in two cases:

- Consumer uses synchronous-style calls

Any error that is `throw`n from within a `get()` or `set()` function can be caught and handled.

- Consumer uses event-based calls

Any error that crops up can be emitted as an `error` event; however extra information is needed to associate it with a request. Exposing the sequence number within `error` events as well as returning it from calls to `get()/set()` would probably be the easiest way to handle that.

This is all making me depressed about the current API design. Maybe that'll be the motivation I need to actually knock out https://github.com/TuyaAPI/driver...

Reply:
For now since it's possible for `data` events to contain strings anyways it should be fine.